fearless/Common/Model/ChainRegistry/ChainModel.swift

```diff
@@ -35,12 +50,51 @@ class ChainModel: Codable {
     let addressPrefix: UInt16
     let types: TypesSettings?
     let icon: URL?
-    let options: [ChainOptions]?
+    let options: [ChainOptions]
     let externalApi: ExternalApiSet?
     let selectedNode: ChainNodeModel?
-    let customNodes: Set<ChainNodeModel>?
+    let customNodes: Set<ChainNodeModel>
     let iosMinAppVersion: String?
 
+    enum DecodingError: Error {
+        case missingAssets
+        case missingNodes
+    }
+
+    required init(from decoder: Decoder) throws {
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+
+        // Rules for app stable work
+        // 1. Optional unwrap with "try?"
+        // 2. Strict fields unwrap with "try"
+        // 3. Sets/URLs unwrap with decodeOptionalArray with empty set
+        // 4. Throw error if something is wrong with "DecodingError", so ignore chain in app
+
+        chainId = try container.decode(Id.self, forKey: .chainId)
+        parentId = try? container.decode(Id.self, forKey: .parentId)
+        name = try container.decode(String.self, forKey: .name)
+        assets = container.decodeOptionalArray([ChainAssetModel].self, forKey: .assets).toSet()
+        nodes = container.decodeOptionalArray([ChainNodeModel].self, forKey: .nodes).toSet()
+        addressPrefix = try container.decode(UInt16.self, forKey: .addressPrefix)
+        types = try? container.decode(TypesSettings.self, forKey: .types)
+        icon = try? container.decode(URL.self, forKey: .icon)
+        options = container.decodeOptionalArray([ChainOptions].self, forKey: .options)
+        externalApi = try? container.decode(ExternalApiSet.self, forKey: .externalApi)
+        selectedNode = try? container.decode(ChainNodeModel.self, forKey: .selectedNode)
+        customNodes = container.decodeOptionalArray([ChainNodeModel].self, forKey: .customNodes).toSet()
+        iosMinAppVersion = try? container.decode(String.self, forKey: .iosMinAppVersion)
+
+        if assets.isEmpty {
+            throw DecodingError.missingAssets
+        }
+
+        if nodes.isEmpty {
+            throw DecodingError.missingNodes
+        }
+
+        // TODO: why types optional?
```

Comment:
Already done?

Reply:
Yeah, types is optional because it's backward-compatibility for old runtime metadatas, when remote types file not used